release/models/hashing/openconfig-hashing.yang

```diff
@@ -0,0 +1,238 @@
+
+    typedef hash-field-type-ipv6 {
+        type enumeration {
+            enum NONE {
+                description "Do not use any IPv6 fields in the
+                    calculation of the hash.";
+            }
+            enum SRC_ADDRESS {
+                description "Use only the IPv6 source address in the
+                    calculation of the hash.";
+            }
+            enum DST_ADDRESS {
+                description "Use only the IPv6 destination address in the
+                    calculation of the hash.";
+            }
+            enum BOTH {
+                description "Use both the IPv6 source and
+                    destination addresses in the calculation of the
+                    hash.";
+            }
+            enum TARGET_DEFINED {
+                description "Use the default fields that are used by
+                    the device for calculating the hash.";
+            }
+            enum FLOW_LABEL {
+                description "Use the flow label in the IPv6 header
+                    to calculate the hash.";
+            }
+        }
+        description "The IPv6 fields that should be used to
+            compute the hash.";
+    }
+
+    typedef hash-field-type-transport {
+        type enumeration {
+            enum NONE {
+                description "Do not use any transport fields in the
+                    calculation of the hash.";
+            }
+            enum SRC_PORT {
+                description "Use only the source port from the
+                    transport header in the calculation of the hash.";
+            }
+            enum DST_PORT {
+                description "Use only the destination port from the
+                    transport header in the calculation of the hash.";
+            }
+            enum BOTH {
+                description "Use both the source and
+                    destination ports from the transport header in
+                    the calculation of the hash.";
+            }
+            enum TARGET_DEFINED {
+                description "Use the default fields that are used by
+                    the device for calculating the hash.";
+            }
+        }
+        description "The trasport layer fields that should be used to
+            compute the hash.";
+    }
+
+    grouping hashing-inputs {
+        description
+            "Top level container for inputs to be used for hashing policies.";
+
+        leaf ipv4 {
+            type hash-field-type-ipv4;
+            description "The IPv4 fields that should be used to
+                compute the hash.";
+        }
+        leaf ipv6 {
+            type hash-field-type-ipv6;
+            description "The IPv6 fields that should be used to
+                compute the hash.";
+        }
+        leaf transport-ports {
+            type hash-field-type-transport;
+            description "The transport layer fields that should be used to
+                compute the hash.";
+        }
```

Comment:
I would like to see additionaly:
IPv4_in_IPv4 and IPv6-in-IPv4 where  each lement is again list of outher and inner fields:
ipv4_ipv4/outherfields/outherfield type hash-field-type-ipv4
ipv4_ipv4/innerfields/innerfield type hash-field-type-ipv4
ipv6_ipv4/outherfields/outherfield/ipv4 hash-field-type-ipv6
ipv6_ipv4/innerfields/innerfield/ipv6 hash-field-type-ipv6

---------
Similiarly fot  ip-over-GRE over IPv4/IPv6,
grev4/outherfields/outherfield type hash-field-type-ipv4
grev4/payloads/payload enum(ipv4 ipv6 mpls Eth) // use fields deviend for basic v4,v6,mpls,eth depending on GRE protocol. if not on list 
grev6/outherfields/outherfield type hash-field-type-ipv6
grev6/payloads/payload enum(ipv4 ipv6 mpls Eth) // use fields deviend for basic v4,v6,mpls,eth depending on GRE protocol. if not on list 

--------
Also for MPLS
mpls/labels - number of labels to use
mpls/base enum bottom XOR top default top
mpls/payloads/payload enum (ipv4 ipv6 PWE3 ipv4_grev4 ipv6_grev6 ipv4_ipv4 ipv6_ipv4 ...)

----------
Also Eth- usefull for PWE3 services or L2 switching devices
eth/fields/fild enum (srcMAC,dstMAC,Vlan-ID, EtherType/payload)
eth/payloads/payload enum ((ipv4 ipv6 PWE3 ipv4_grev4 ipv6_grev6 ipv4_ipv4 ipv6_ipv4 ...)

Reply:
For this PR we will only introduce the simple field selection which is targeted for gPINS release and is at least partially supported by the referenced implementations in this PR.

It is possible to  add additional encapsulations in the future without breaking changes to the model by adding additional leafs to this hashing-inputs grouping.  The additional leafs may be added to the public OC model in the future, or they can be vendor augments.

Since there are many constraints in hardware as to which encapsulations may be supported, especially for more complex use cases, we should leave open the option that vendors are expected to augment the model with the specific scenarios they support.  These are often custom or one-off items that are not generally available and may also not be formally documented, which is a good reason to use a vendor provided augment.  These 'custom' options may also not be very 'configurable'.  For example, hardware may support only using a subset of the bits of some fields (ipv6 comes to mind).  Which bits are used may not be configurable or even exposed to the user through the model, config or even the SDK.  Using a vendor augment, a vendor can expose the ability to turn on or off these capabilities.